File: 45_isomorphic_strings.cpp
#include <bits/stdc++.h>
using namespace std;

/*
Task 45: Check if two strings are isomorphic.
Reads two lines and prints YES/NO.
*/

bool isIsomorphic(const string& s, const string& t)
{
    if (s.size() != t.size())
        return false;
    vector<int> ms(256, -1), mt(256, -1);
    for (int i = 0; i < s.size(); ++i)
    {
        unsigned char a = s[i], b = t[i];
        if (ms[a] != mt[b])
            return false;
        ms[a] = mt[b] = i + 1;
    }
    return true;
}

int main()
{
    string a, b;
    if (!getline(cin, a))
        return 0;
    if (!getline(cin, b))
        return 0;
    cout << (isIsomorphic(a, b) ? "YES" : "NO") << "\n";
}

// Time Complexity: O(n)
// Space Complexity: O(1)

File: 45_isomorphic_strings.py
"""
Task 45: Check if two strings are isomorphic.
Reads two lines and prints YES/NO.
"""
import sys

def is_isomorphic(s: str, t: str) -> bool:
    if len(s) != len(t): return False
    m1 = {}
    m2 = {}
    for i,ch in enumerate(s):
        if ch in m1 and m1[ch] != t[i]: return False
        if t[i] in m2 and m2[t[i]] != ch: return False
        m1[ch] = t[i]
        m2[t[i]] = ch
    return True

if __name__ == '__main__':
    a = sys.stdin.readline().rstrip('\n')
    b = sys.stdin.readline().rstrip('\n')
    print('YES' if is_isomorphic(a,b) else 'NO')

# Time Complexity: O(n)
# Space Complexity: O(k)
